### ISSUE TITLE
Extending Map is unweildy

### ISSUE THREAD

**Issue by jbccollins** [PaulLeCam/react-leaflet#268]
### Expected behavior

Extending a the base Map component is not as easy as extending other things. I have a use case where I want to use a custom Map that overrides some of the leaflet functions. I would like to be able to just have my own CustomMap.js file that looks like this:

```javascript
import 'custom-leaflet-map'; // Defines my custom map using L.Map.extend
import { Map } from 'react-leaflet';
import L from 'leaflet';
import _ from 'lodash';
const { isUndefined, omit } = _;

export default class CustomMap extends Map {
  componentDidMount () {
    const props = omit(this.props, ['children', 'className', 'id', 'style']);
    this.leafletElement = L.customMap(this.container, props);
    super.componentDidMount();
    this.setState({map: this.leafletElement});
    if (!isUndefined(props.bounds)) {
      this.leafletElement.fitBounds(props.bounds, props.boundsOptions);
    }
  }
}
```

so that all I have to override is the componentDidMount function. However, since you can't reinitialize a leaflet map once it already exists this will throw an error when trying to intitialize my custom map. 
The way I see it my only alternative is to directly call 'this.bindLeafletEvents(this._leafletEvents);' instead of super.componentDidMount() which right now will work but will be a pain to debug if that function is ever updated some time in the future.

**Comment by PaulLeCam** [comment]
I think the main problem is that creating and updating the Leaflet element is done directly in the React lifecycle handlers, while it should be abstracted from it.
I started to refactor the library in [the split-logic branch](https://github.com/PaulLeCam/react-leaflet/tree/split-logic) to extract the element creation as `this.leafletElement = createLeafletElement(props)` and update as `updateLeafletElement(fromProps, toProps)`, I still need to run more tests around it but if you want to try out this branch, it should be stable enough.

**Comment by jbccollins** [comment]
@PaulLeCam Thanks for the quick reply. I'll take a look at that branch.

**Comment by jbccollins** [comment]
@PaulLeCam Ok. I checked out that branch and It's much better. I can get what I want with just:

```javascript
export default class CustomMap extends Map {
  createLeafletElement (props: Object): Object {
    return L.customMap(this.container, props);
  }
}
```

Any idea when this branch will be merged and published?

**Comment by PaulLeCam** [comment]
I just published the changes from this branch in v1.1.0
